tests/test_setup_kubernetes_job.py

```diff
@@ -56,40 +116,103 @@ def test_main():
         assert e.value.code == 1
 
 
-def test_setup_kube_deployment_invalid_job_name():
+def test_get_service_instances_with_valid_names():
+    mock_service_instances = ["kurupt.f_m"]
+    ret = get_service_instances_with_valid_names(
+        service_instances=mock_service_instances
+    )
+
+    assert ret == [("kurupt", "f_m", None, None)]
+
+
+def test_main_invalid_job_name():
     with mock.patch(
+        "paasta_tools.setup_kubernetes_job.parse_args", autospec=True
+    ) as mock_parse_args, mock.patch(
+        "paasta_tools.setup_kubernetes_job.KubeClient", autospec=True
+    ), mock.patch(
+        "paasta_tools.setup_kubernetes_job.metrics_lib.get_metrics_interface",
+        autospec=True,
+    ), mock.patch(
+        "paasta_tools.setup_kubernetes_job.load_system_paasta_config", autospec=True
+    ), mock.patch(
         "paasta_tools.setup_kubernetes_job.create_application_object", autospec=True
-    ) as mock_create_application_object, mock.patch(
-        "paasta_tools.setup_kubernetes_job.list_all_deployments", autospec=True
-    ) as mock_list_all_deployments, mock.patch(
-        "paasta_tools.setup_kubernetes_job.log", autospec=True
-    ):
-        mock_client = mock.Mock()
-        mock_list_all_deployments.return_value = [
-            KubeDeployment(
-                service="kurupt",
-                instance="f_m",
-                git_sha="",
-                image_version=None,
-                config_sha="",
-                replicas=0,
-            )
+    ) as mock_create_application_object:
+        mock_parse_args.return_value.cluster = "fake_cluster"
+        mock_parse_args.return_value.soa_dir = "/etc/fake"
+        mock_parse_args.return_value.service_instance_list = ["kuruptf_m"]
+        with raises(SystemExit) as e:
+            main()
+        assert mock_create_application_object.call_count == 0
+        assert e.value.code == 1
+
+
+def test_get_kubernetes_deployment_config():
+    with mock.patch(
+        "paasta_tools.setup_kubernetes_job.get_service_instances_with_valid_names",
+        autospec=True,
+    ) as mock_get_service_instances_with_valid_names, mock.patch(
+        "paasta_tools.setup_kubernetes_job.load_kubernetes_service_config_no_cache",
+        autospec=True,
+    ) as mock_load_kubernetes_service_config_no_cache:
+
+        mock_get_service_instances_with_valid_names = [
+            ("kurupt", "instance", None, None)
         ]
-        mock_service_instances = ["kuruptf_m"]
-        setup_kube_deployments(
-            kube_client=mock_client,
-            service_instances=mock_service_instances,
+        mock_load_kubernetes_service_config_no_cache.side_effect = (
+            NoDeploymentsAvailable
+        )
+        ret = get_kubernetes_deployment_config(
+            service_instances_with_valid_names=mock_get_service_instances_with_valid_names,
```

Comment:
Should you be calling `mock_get_service_instances_with_valid_names()` instead of passing the mock object directly? I could be misinterpreting how this test is supposed to work.

Reply:
In this test, I wanted to test the different return values of ``get_kubernetes_deployment_config`` function, so first I tested the case where we have ``NoDeploymentsAvailable``. I wanted to see if it will return ``[(True, None)]``. Then I tested the case where we have ``NoConfigurationForServiceError`` and checked if the return value was ``[(False, None)]``. Lastly, I wanted to test if I passed a ``KubernetesDeploymentConfig`` would my return value be ``(True, app)``. I already assumed that ``mock_get_service_instances_with_valid_names()`` is valid and focused on the different return values I can get from this `get_kubernetes_deployment_config`` function

Reply:
I tested ``get_service_instances_with_valid_names`` in this test:
```
def test_get_service_instances_with_valid_names():
    mock_service_instances = ["kurupt.f_m"]
    ret = get_service_instances_with_valid_names(
        service_instances=mock_service_instances
    )

    assert ret == [("kurupt", "f_m", None, None)]
```

Reply:
Please let me know if this test makes sense or if there is a better way of testing ``test_get_kubernetes_deployment_config`` and ``test_get_service_instances_with_valid_names()``

Reply:
Oh, I think I understand why I was confused before. `mock_get_service_instances_with_valid_names` is defined twice in this test -- it gets defined as a mock object by `mock.patch(...) as mock_get_service_instances_with_valid_names` on lines 161-164, and then re-defined as a list on line 169-171. The list is what you're passing to the function under test. I hadn't noticed the re-definition before and assumed you were passing a mock object which was pretending to be a function, not a list.

Do you even need to mock.patch `get_service_instances_with_valid_names` in this test? I think `get_service_instances_with_valid_names` is only called by the `main` method in setup_kubernetes_job, which this test doesn't call. The test passes if I remove it:


```
krall@dev56-uswest1adevc:~/pg/paasta (u/emanelsabban/PAASTA-17775) $ git diff
diff --git a/tests/test_setup_kubernetes_job.py b/tests/test_setup_kubernetes_job.py
index 9dfceef5d..57b7d22c0 100644
--- a/tests/test_setup_kubernetes_job.py
+++ b/tests/test_setup_kubernetes_job.py
@@ -159,9 +159,6 @@ def test_main_invalid_job_name():

 def test_get_kubernetes_deployment_config():
     with mock.patch(
-        "paasta_tools.setup_kubernetes_job.get_service_instances_with_valid_names",
-        autospec=True,
-    ) as mock_get_service_instances_with_valid_names, mock.patch(
         "paasta_tools.setup_kubernetes_job.load_kubernetes_service_config_no_cache",
         autospec=True,
     ) as mock_load_kubernetes_service_config_no_cache:
krall@dev56-uswest1adevc:~/pg/paasta (u/emanelsabban/PAASTA-17775) $ make test
...
  tests-yelpy: commands succeeded
  congratulations :)
krall@dev56-uswest1adevc:~/pg/paasta (u/emanelsabban/PAASTA-17775) $
```

Reply:
yes, you are right , will remove the mock object